### PR TITLE
optimize: RM TM startup connect server fail fast

### LINKED ISSUE #5484: support active check of service availability at client startup
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
类似dubbo的服务检查,如果没有发现该服务,或者无法与该服务的provider建立连接,将抛出异常阻止应用启动
<dubbo:reference interface="x.x.x.xxxx" check="true" />

image
现在一启动连不上server,也只是打个error日志,无法中断一样启动,特别是rm的连接是异步10s后建立,如果在生产可能会由于配置等问题,造成故障

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- Seata version:  
- OS :
- Others:

### PR DESCRIPTION
**Pull request by jsbxyyx**
- [ ] I have registered the PR [changes](../changes).

### Ⅰ. Describe what this PR did
fixes: #5484

### CODE REVIEW

File: core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java [hunk 1]
```diff
@@ -181,12 +181,21 @@ void reconnect(List<String> availList, String transactionServiceGroup) {
...
+                            failedMap.values().stream().map(Throwable::getMessage).collect(Collectors.toSet()));
                 } else if (LOGGER.isDebugEnabled()) {
                     failedMap.forEach((key, value) -> {
-                        LOGGER.error("{} can not connect to {} cause:{} trace information:{}", FrameworkErrorCode.NetConnect.getErrCode(), key, value.getMessage(), value);
+                        LOGGER.error("{} can not connect to {} cause:{} trace information:{}",
+                                FrameworkErrorCode.NetConnect.getErrCode(), key, value.getMessage(), value);
                     });
                 }
+                boolean failFast = NettyClientConfig.isEnableClientChannelCheckFailFast();
+                boolean main = Thread.currentThread().getId() == 1L;
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java]: Is threadID=1 always the main/initial thread ?

Consider an external Tomcat container. Is the application running on the main thread, and is the thread ID 1?

File: core/src/main/java/io/seata/core/rpc/netty/AbstractNettyRemotingClient.java [hunk 2]
```diff
@@ -106,7 +106,13 @@ public abstract class AbstractNettyRemotingClient extends AbstractNettyRemoting
 
     @Override
     public void init() {
```
Comment (wt-better) [inline core/src/main/java/io/seata/core/rpc/netty/AbstractNettyRemotingClient.java:106]: if fastfail, Is it inappropriate to still connect tc in an asynchronous thread?

**Review by funky-eyes — DISMISSED** [review]
LGTM

File: core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java [hunk 3]
```diff
@@ -161,13 +161,67 @@ void destroyChannel(String serverAddress, Channel channel) {
...
+
+            if (!(registryService instanceof FileRegistryServiceImpl)) {
+                LOGGER.error("no available service found in cluster '{}', please make sure registry config correct and keep your seata server running", clusterName);
+            }
+            return;
+        }
+        try {
+            doReconnect(availList, transactionServiceGroup);
+        } catch (Exception e) {
+            if (failFast) throw e;
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java]: if failfast=false or schedule reconnect, there is no print when the exception is thrown? As if it didn't happen?

**Comment by laywin** [comment]
1. may be fast fail is there are no available server from registry throw exception when start up, not there are available server but some can't connect. the check process can keep the available server list clear.
2. In addition, i think the available server map can hook into registry, when the server not online, available server map can first time get the notify message not depend on 10s to polling.

File: core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java (hunk 3 above)
Comment (jsbxyyx) [reply core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java]: done

**Comment by slievrly** [comment]
> 1. may be fast fail is there are no available server from registry throw exception when start up, not there are available server but some can't connect. the check process can keep the available server list clear.
> 2. In addition, i think the available server map can hook into registry, when the server not online, available server map can first time get the notify message not depend on 10s to polling.

The server online/offline notifications in the registry center are based on a subscription mechanism. The "10s" mentioned here refers to the interval for checking the availability of servers. The registry center list displays the nodes that are deemed normal, but it doesn't guarantee that the clients can necessarily connect to them.

File: core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java [hunk 4]
```diff
@@ -161,13 +161,70 @@ void destroyChannel(String serverAddress, Channel channel) {
...
+            }
+            return;
+        }
+        try {
+            doReconnect(availList, transactionServiceGroup);
+        } catch (Exception e) {
+            if (failFast) {
+                throw e;
+            }
+            LOGGER.error("connect TC failed. {}", e.getMessage(), e);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java]: TC is a logical role, and I think it is more suitable to write as "Server". The terminology needs to be standardized, please refer to other logs.

**Comment by laywin** [comment]
> > 1. may be fast fail is there are no available server from registry throw exception when start up, not there are available server but some can't connect. the check process can keep the available server list clear.
> > 2. In addition, i think the available server map can hook into registry, when the server not online, available server map can first time get the notify message not depend on 10s to polling.
> 
> The server online/offline notifications in the registry center are based on a subscription mechanism. The "10s" mentioned here refers to the interval for checking the availability of servers. The registry center list displays the nodes that are deemed normal, but it doesn't guarantee that the clients can necessarily connect to them.

Yes, I know the current situation. What I mean is that the list after check can also be referenced in the process of subscription monitoring to determine whether the node can be deleted at the first time when it goes offline, and 10s check is still in place.

是的，我知道现在的情况，我的意思是check 后的列表一样可以引用到订阅监听的过程中去，用来判断当节点下线时，能够第一时间删除掉， 10s check还在.

File: core/src/main/java/io/seata/core/rpc/netty/RmNettyRemotingClient.java [hunk 5]
```diff
@@ -82,7 +82,8 @@ public void init() {
             if (resourceManager != null
                     && !resourceManager.getManagedResources().isEmpty()
                     && StringUtils.isNotBlank(transactionServiceGroup)) {
-                getClientChannelManager().reconnect(transactionServiceGroup);
+                boolean failFast = NettyClientConfig.isEnableClientChannelCheckFailFast();
```
Comment (wt-better) [inline core/src/main/java/io/seata/core/rpc/netty/RmNettyRemotingClient.java]: I understand that fastfail should be an attribute on the application side, not a netty attribute.

Comment (jsbxyyx) [reply core/src/main/java/io/seata/core/rpc/netty/RmNettyRemotingClient.java]: done

File: common/src/main/java/io/seata/common/DefaultValues.java [hunk 6]
```diff
@@ -60,6 +60,8 @@ public interface DefaultValues {
     boolean DEFAULT_ENABLE_RM_CLIENT_BATCH_SEND_REQUEST = true;
     boolean DEFAULT_ENABLE_TC_SERVER_BATCH_SEND_RESPONSE = false;
 
+    boolean DEFAULT_CLIENT_CHANNEL_CHECK_FAIL_FAST = false;
```
Comment (wt-better) [inline common/src/main/java/io/seata/common/DefaultValues.java]: `FAIL_FAST` defaultValue is `true` or better ? 

Only saga mode is possible without depend serverList

File: test/src/test/resources/file.conf [hunk 7]
```diff
@@ -46,8 +46,12 @@ transport {
   }
   serialization = "seata"
   compressor = "none"
+
```
Comment (wt-better) [inline test/src/test/resources/file.conf:49]: file.conf default value also to 'true',  ensure consistency

Comment (jsbxyyx) [reply test/src/test/resources/file.conf:49]: Will cause unit tests to fail

File: changes/en-us/2.x.md [hunk 8]
```diff
@@ -39,13 +39,15 @@ Add changes here for all PR submitted to the 2.x branch.
 - [[#6184](https://github.com/apache/incubator-seata/pull/6184)] update NOTICE file
 - [[#6192](https://github.com/apache/incubator-seata/pull/6192)] remove the useless file
 - [[#6194](https://github.com/apache/incubator-seata/pull/6194)] fix asf.yaml parse error
+- [[#6116](https://github.com/seata/seata/pull/6116)] rewrite NettyPoolKey's hashcode and equals to fix duplicate construction of channel object pools
```
Comment (funky-eyes) [inline changes/en-us/2.x.md]:
```suggestion
```

File: core/src/test/java/io/seata/core/rpc/netty/TmNettyClientTest.java [hunk 9]
```diff
@@ -123,6 +135,25 @@ public void setApplicationId() throws Exception {
...
+    @Test
+    public void testCheckFailFast() throws Exception {
+        TmNettyRemotingClient.getInstance().destroy();
+        TmNettyRemotingClient tmClient = TmNettyRemotingClient.getInstance("fail_fast", "default_tx_group");
+        System.setProperty("file.listener.enabled", "true");
+        ConfigurationCache.addConfigListener(ConfigurationKeys.ENABLE_TM_CLIENT_CHANNEL_CHECK_FAIL_FAST,
+            event -> logger.info("dataId:{}, value: {}, oldValue: {}", event.getDataId(), event.getNewValue(),
+                event.getOldValue()));
+        System.setProperty(ConfigurationKeys.ENABLE_TM_CLIENT_CHANNEL_CHECK_FAIL_FAST, "true");
+        Thread.sleep(2000);
```
Comment (funky-eyes) [inline core/src/test/java/io/seata/core/rpc/netty/TmNettyClientTest.java:153]:
```suggestion
```

File: core/src/test/java/io/seata/core/rpc/netty/RmNettyClientTest.java [hunk 10]
```diff
@@ -47,6 +78,24 @@ public void assertGetInstanceAfterDestroy() {
...
+        Map<String, Resource> resourceMap = new HashMap<>();
+        resourceMap.put("jdbc:xx://localhost/test", mockResource);
+        Mockito.when(resourceManager.getManagedResources()).thenReturn(resourceMap);
+        newClient.setResourceManager(resourceManager);
+        System.setProperty("file.listener.enabled", "true");
+        ConfigurationCache.addConfigListener(ConfigurationKeys.ENABLE_RM_CLIENT_CHANNEL_CHECK_FAIL_FAST,
+            event -> logger.info("dataId:{}, value: {}, oldValue: {}", event.getDataId(), event.getNewValue(),
+                event.getOldValue()));
+        System.setProperty(ConfigurationKeys.ENABLE_RM_CLIENT_CHANNEL_CHECK_FAIL_FAST, "true");
+        Assertions.assertThrows(FrameworkException.class, newClient::init);
```
Comment (funky-eyes) [inline core/src/test/java/io/seata/core/rpc/netty/RmNettyClientTest.java:98]:
```suggestion
        Thread.sleep(2000);
        Assertions.assertThrows(FrameworkException.class, newClient::init);
```

File: core/src/test/java/io/seata/core/rpc/netty/TmNettyClientTest.java [hunk 11]
```diff
@@ -123,6 +135,24 @@ public void setApplicationId() throws Exception {
...
+    @Test
+    public void testCheckFailFast() throws Exception {
+        TmNettyRemotingClient.getInstance().destroy();
+        TmNettyRemotingClient tmClient = TmNettyRemotingClient.getInstance("fail_fast", "default_tx_group");
+        System.setProperty("file.listener.enabled", "true");
+        ConfigurationCache.addConfigListener(ConfigurationKeys.ENABLE_TM_CLIENT_CHANNEL_CHECK_FAIL_FAST,
+            event -> logger.info("dataId:{}, value: {}, oldValue: {}", event.getDataId(), event.getNewValue(),
+                event.getOldValue()));
+        System.setProperty(ConfigurationKeys.ENABLE_TM_CLIENT_CHANNEL_CHECK_FAIL_FAST, "true");
+        Assertions.assertThrows(FrameworkException.class, tmClient::init);
```
Comment (funky-eyes) [inline core/src/test/java/io/seata/core/rpc/netty/TmNettyClientTest.java:154]:
```suggestion
        Thread.sleep(2000);
        Assertions.assertThrows(FrameworkException.class, tmClient::init);
```

**Review by funky-eyes — APPROVED** [review]
LGTM

**Pull request merged**